File: typescript/packages/server/src/version/restore.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  applyStateDict,
  toStateDict,
  type Workspace as CoreWorkspace,
  type WorkspaceStateDict,
} from '@struktoai/mirage-core'
import { readVersion, resolveRef } from './api.ts'
import { grantWidenings, sessionsDiff, type GrantWidening } from './stateDiff.ts'
import { toState } from './stateTree.ts'
import type { VersionStore } from './store.ts'

type AnyDict = Record<string, unknown>

const PLANES = ['files', 'sessions', 'namespace', 'history'] as const
type Plane = (typeof PLANES)[number]

export interface RestoreReport {
  version: string
  planes: Plane[]
  paths: string[] | null
  grant_widenings: GrantWidening[]
}

export interface RestoreOptions {
  paths?: string[]
  planes?: Plane[]
}

function normPath(p: string): string {
  let start = 0
  let end = p.length
  while (start < end && p[start] === '/') start++
  while (end > start && p[end - 1] === '/') end--
  return `/${p.slice(start, end)}`
}

function selectedFile(path: string, wanted: string[]): boolean {
  const p = normPath(path)
  return wanted.some((w) => {
    const want = normPath(w)
    return p === want || p.startsWith(`${want}/`)
  })
}

function mergeMountFiles(
  liveMount: AnyDict,
  targetMount: AnyDict,
  prefix: string,
  wanted: string[],
): void {
  const liveState = liveMount.resource_state as AnyDict
  const targetFiles =
    ((targetMount.resource_state as AnyDict).files as Record<string, Uint8Array> | undefined) ?? {}
  const files = { ...((liveState.files as Record<string, Uint8Array> | undefined) ?? {}) }
  const drop = new Set<string>()
  const base = prefix.replace(/\/+$/, '')
  for (const rel of new Set([...Object.keys(files), ...Object.keys(targetFiles)])) {
    if (!selectedFile(`${base}${rel}`, wanted)) continue
    const target = targetFiles[rel]
    if (target !== undefined) files[rel] = target
    else drop.add(rel)
  }
  liveState.files = Object.fromEntries(Object.entries(files).filter(([rel]) => !drop.has(rel)))
}

/**
 * Surgical restore: whole world, chosen planes, or chosen paths.
 *
 * Scope rules: no options = the whole world (checkout semantics);
 * `planes` picks a subset and leaves the live state of the others
 * untouched; `paths` restores only the matching files (a path selects
 * itself or its subtree) and implies the files plane alone. Restoring
 * sessions re-applies their mount grants, so the report carries
 * `grant_widenings`: every grant the restore widened relative to the
 * live state, surfaced, never silent. Mirrors the Python restore.
 */
export async function restore(
  store: VersionStore,
  ws: CoreWorkspace,
  ref: string,
  options: RestoreOptions = {},
): Promise<RestoreReport> {
  const { paths, planes } = options
  if (paths !== undefined && planes !== undefined) {
    throw new Error('restore takes paths or planes, not both')
  }
  for (const plane of planes ?? []) {
    if (!PLANES.includes(plane)) {
      throw new Error(`unknown plane '${plane as string}'; expected one of ${PLANES.join(', ')}`)
    }
  }
  const version = await resolveRef(store, ref)
  const { entries, meta } = await readVersion(store, version)
  const target = toState(entries, meta) as unknown as AnyDict
  const live = (await toStateDict(ws)) as unknown as AnyDict
  const fallback: readonly Plane[] = paths !== undefined ? ['files'] : PLANES
  const selected = new Set<Plane>(planes ?? fallback)

  let widenings: GrantWidening[] = []
  if (selected.has('sessions')) {
    const delta = sessionsDiff(
      (live.sessions as AnyDict[] | undefined) ?? [],
      (target.sessions as AnyDict[] | undefined) ?? [],
    )
    widenings = grantWidenings(delta)
  }

  const merged: AnyDict = { ...target }
  if (!selected.has('sessions')) {
    merged.sessions = (live.sessions as AnyDict[] | undefined) ?? []
    merged.default_session_id = live.default_session_id
  }
  if (!selected.has('namespace')) merged.nodes = (live.nodes as AnyDict | undefined) ?? {}
  if (!selected.has('history')) merged.history = (live.history as AnyDict[] | undefined) ?? []
  if (!selected.has('files')) {
    merged.mounts = (live.mounts as AnyDict[] | undefined) ?? []
  } else if (paths !== undefined) {
    const targetByPrefix = new Map(
      ((target.mounts as AnyDict[] | undefined) ?? []).map((m) => [m.prefix as string, m]),
    )
    merged.mounts = (live.mounts as AnyDict[] | undefined) ?? []
    for (const mount of merged.mounts as AnyDict[]) {
      const prefix = mount.prefix as string
      const targetMount = targetByPrefix.get(prefix)
      if (targetMount !== undefined) mergeMountFiles(mount, targetMount, prefix, paths)
    }
  }

  const cache = ws.cache as { clear?: () => Promise<void> }
  if (typeof cache.clear === 'function') await cache.clear()
  await applyStateDict(ws, merged as unknown as WorkspaceStateDict)
  return {
    version,
    planes: [...selected].sort(),
    paths: paths ?? null,
    grant_widenings: widenings,
  }
}

File: python/mirage/server/version/restore.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.server.version.api import read_version, resolve_ref
from mirage.server.version.state_tree import CATEGORIES, to_state
from mirage.server.version.store import VersionStore
from mirage.types import DriftPolicy, MountKey, ResourceStateKey, StateKey
from mirage.utils.path import norm
from mirage.workspace.snapshot import (apply_state_dict, install_fingerprints,
                                       to_state_dict)


def _selected_file(path: str, wanted: list[str]) -> bool:
    p = norm(path)
    for want in wanted:
        w = norm(want)
        if p == w or p.startswith(w + "/"):
            return True
    return False


def _merge_mount_files(live_mount: dict[str, Any], target_mount: dict[str,
                                                                      Any],
                       prefix: str, wanted: list[str]) -> None:
    live_state = live_mount[MountKey.RESOURCE_STATE]
    target_files = target_mount[MountKey.RESOURCE_STATE].get(
        ResourceStateKey.FILES, {})
    files = dict(live_state.get(ResourceStateKey.FILES, {}))
    base = prefix.rstrip("/")
    for rel in set(files) | set(target_files):
        full = f"{base}{rel}"
        if not _selected_file(full, wanted):
            continue
        if rel in target_files:
            files[rel] = target_files[rel]
        else:
            files.pop(rel, None)
    live_state[ResourceStateKey.FILES] = files


async def restore(
        store: VersionStore,
        ws,
        ref,
        *,
        paths: list[str] | None = None,
        categories: list[str] | None = None,
        drift_policy: DriftPolicy = DriftPolicy.STRICT) -> dict[str, Any]:
    """Surgical restore: whole world, chosen categories, or chosen paths.

    Scope rules: no arguments = the whole world (checkout semantics);
    ``categories`` picks a subset of files/sessions/namespace/history and
    leaves the live state of the others untouched; ``paths`` restores
    only the matching files (a path selects itself or its subtree) and
    implies the files category alone. Restoring sessions re-applies
    their mount grants exactly like any other state; compare two
    versions with :func:`state_diff` to see grant changes up front.
    """
    if paths is not None and categories is not None:
        raise ValueError("restore takes paths= or categories=, not both")
    for category in categories or []:
        if category not in CATEGORIES:
            raise ValueError(f"unknown category {category!r}; expected one of "
                             f"{', '.join(CATEGORIES)}")
    version = await resolve_ref(store, ref)
    entries, meta = await read_version(store, version)
    target = to_state(entries, meta)
    live = await to_state_dict(ws)
    selected = (set(categories) if categories is not None else
                {"files"} if paths is not None else set(CATEGORIES))

    merged = dict(target)
    if "sessions" not in selected:
        merged[StateKey.SESSIONS] = live.get(StateKey.SESSIONS) or []
        merged[StateKey.DEFAULT_SESSION_ID] = live.get(
            StateKey.DEFAULT_SESSION_ID)
    if "namespace" not in selected:
        merged[StateKey.NODES] = live.get(StateKey.NODES) or {}
    if "history" not in selected:
        merged[StateKey.HISTORY] = live.get(StateKey.HISTORY) or []
    if "files" not in selected:
        merged[StateKey.MOUNTS] = live.get(StateKey.MOUNTS) or []
    elif paths is not None:
        target_by_prefix = {
            m[MountKey.PREFIX]: m
            for m in target.get(StateKey.MOUNTS) or []
        }
        merged[StateKey.MOUNTS] = live.get(StateKey.MOUNTS) or []
        for mount in merged[StateKey.MOUNTS]:
            prefix = mount[MountKey.PREFIX]
            if prefix in target_by_prefix:
                _merge_mount_files(mount, target_by_prefix[prefix], prefix,
                                   paths)

    await ws._cache.clear()
    await apply_state_dict(ws, merged)
    if "files" in selected and paths is None:
        install_fingerprints(ws,
                             target.get(StateKey.FINGERPRINTS) or [],
                             drift_policy)
    return {
        "version": version.decode(),
        "categories": sorted(selected),
        "paths": paths,
    }
